### ISSUE TITLE
 KiCad project switches to local temp copy after export

### ISSUE THREAD

**Issue by wilhelmzeuschner** [30350n/pcb2blender#68]
After an export, the KiCad project "switches" to the temp files created during the export process.
I'm using the File -> save copy command here to demonstrate:
![image](https://github.com/user-attachments/assets/9d359e09-24d0-4298-ba2c-ecd750ce2108)

This is especially evident when reopening KiCad and opening the most recent project:
![image](https://github.com/user-attachments/assets/f3505ef4-5f3a-47fb-a5c9-fa68de7713cc)

If I were to make edits to my project after export, I'll be working in the temp copy and thus will likely lose my changes unintentionally.

**Comment by 30350n** [comment]
Urgh, that's annoying ...
Not sure if I can work around that though ...

**Comment by hraftery** [comment]
Is it caused by this?

https://github.com/30350n/pcb2blender/blob/ee6140001ae42c7bd42cda0b2137750553f0121c/pcb2blender_exporter/export.py#L217

I don't suppose it's a matter of re-opening the board at the end? Something like:

```
pcbnew.LoadBoard(board.filepath);
```

Might be risky, if the user has not saved the current state.

**Comment by 30350n** [comment]
> Is it caused by this?

Yes it is.

> I don't suppose it's a matter of re-opening the board at the end? Something like:
>
>```pcbnew.LoadBoard(board.filepath);```
>
>Might be risky, if the user has not saved the current state.

Yeah, that doesn't sound like a good idea.

To give some more context here on *why* I'm saving the file anyways: The main problem here is that there's no API to access board stackup information.

So I literally just """parse""" (using a bunch of ugly regex)  the `.kicad_pcb` file and extract the data that way:

https://github.com/30350n/pcb2blender/blob/ee6140001ae42c7bd42cda0b2137750553f0121c/pcb2blender_exporter/export.py#L213-L236

And to ensure that the data in the file is up to date, I temporarily save the file to a temp directory.

I don't really want to waste any time looking into this until 9.0 comes out, because there'll be a new python API (which might or might not make all of this much easier).

**Comment by 30350n** [comment]
The current behavior with Saving is arguably worse/more dangerous than not saving I guess, but reimplementing this without saving is a whole another can of worms again so yeah, ...

**Comment by hraftery** [comment]
Yeah fair enough. Thanks for the explanation.

I guess the root cause is that a programmatic board save affects the user's project history, which is a bit of a bummer. But it is what it is.

Hang on, isn't that exactly what [aSkipSettings](https://docs.kicad.org/doxygen-python-7.0/namespacepcbnew.html#a7dd1ee694d047fd271954eb5421c060c) is for? It seems that invokes [SETTINGS_MANAGER](https://docs.kicad.org/doxygen/classSETTINGS__MANAGER.html#a18b71f3bf49bf891fccaf218800d4981) which:

> Sets the currently loaded project path and saves it

So maybe just:

```pcbnew.SaveBoard(str(tmp_path), board, true)```

would do?

**Comment by 30350n** [comment]
Hmmm, first thought was that that looks promising, but looking at [this](https://github.com/KiCad/kicad-source-mirror/blob/c3fb1445c741c465d8528fc1af24dba51700fb20/pcbnew/python/scripting/pcbnew_scripting_helpers.cpp#L323-L351), it looks like `skipSettings` only affects saving a reference to the `.kicad_pcb` file, to the `kicad_pro` file. 🤔 

If anyone want to test this though, feel free to do (I'm not really sure myself, but I lack time to test this atm).

**Comment by hraftery** [comment]
Alright, you nerd sniped me. Results are very promising. Here's my MRE, which is almost entirely lifted from `export.py`:

```
import pcbnew
import shutil, tempfile
from pathlib import Path


def get_tempdir():
    return Path(tempfile.gettempdir()) / "pcb2blender_tmp"

def get_temppath(filename):
    return get_tempdir() / filename

def init_tempdir():
    tempdir = get_tempdir()
    if tempdir.exists():
        try:
            shutil.rmtree(tempdir)
        except OSError:
            try:
                # try to delete all files first
                for file in tempdir.glob("**/*"):
                    if file.is_file():
                        file.unlink()
                shutil.rmtree(tempdir)
            except OSError:
                # if this still doesn't work, fuck it
                return
    tempdir.mkdir()


board = pcbnew.GetBoard()

init_tempdir()
tmp_path = get_temppath("pcb2blender_tmp.kicad_pcb")
pcbnew.SaveBoard(str(tmp_path), board, True) # this is the first line which is not already in export.py


content = tmp_path.read_text(encoding="utf-8")
len(content) # this is the second line which is not already in export.py, and only here for testing
```

If I open a KiCad project, open the `kicad_pcb` file, open the Python console, and then copy and paste this (stanza by stanza because eh), I find:

1. `SaveBoard` returns True
2. `len(content)` shows the exported file exists and can be read
3. quitting and restarting KiCad opens with the project I had open, not the newly exported one. Also, the tmp project does not appear in the recent projects list in the File menu.

I repeated this, but with a dirty file state (I moved some parts on the pcb before running the script) and it behaved as expected - I was prompted to save the changes on quit, and was free to discard them.

I also repeated this without quitting KiCad first, and instead of moving a part, deleted it. I confirmed the content length got smaller, demonstrating that the file is getting generated each time.

Looks like this could be much easier to resolve than it first seemed!

**Comment by 30350n** [comment]
> Alright, you nerd sniped me. Results are very promising.

Haha ^^, thanks for looking into this!

Can you test one more thing?
What happens if you export with a dirty file state (move some stuff around again I guess), then export and finally actually **save** the file when being prompted on exit (or saving manually).

Do the changes get properly saved to the **original** file then?

**Comment by hraftery** [comment]
Good call.

Yes, the original file is updated. The last modified date for the `kicad_pcb` file is updated, and when I re-open KiCad, the change I made is present.

I tried scenarios with changes made before and after running the script, and with invoking "Save" from the File menu or in the quit dialog.

**Comment by 30350n** [comment]
Okay, this sounds really perfect then!

Do you want to create a PR for this or should I just quickly do the changes myself? Should only really involve changing that one line I guess.

**Comment by hraftery** [comment]
Got around to it. Hopefully it's helpful: https://github.com/30350n/pcb2blender/pull/80

**Comment by 30350n** [comment]
Should be fixed in [v2.15](https://github.com/30350n/pcb2blender/releases/tag/v2.15-k8.0-b4.2).

Thanks again everyone for reporting and finding a fix! ❤️
